FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,51 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				if let result = try ManagedCache.cache(in: context) {
+					completion(.found(feed: result.localImages, timestamp: result.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				let managedCache = try ManagedCache.new(in: context)
+				managedCache.timestamp = timestamp
+				managedCache.feed = ManagedFeedImage.imageSet(from: feed, in: context)
+
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				try ManagedCache.delete(in: context)
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
+	}
+
+	private func perform(_ action: @escaping (NSManagedObjectContext) -> Void) {
+		let context = context
+		context.perform {
```

Comment:
Another way of capturing the `context` without capturing `self` inside of the `perform` block is to use a capture list:

```suggestion
		context.perform { [context] in
```

It saves one line and one variable.

No need to make this amendment, this is just for your interest.

Reply:
I like this kind of 'syntactic sugar' thanks :)